### DISCUSSION TITLE
'CUDA_VISIBLE_DEVICES' question

### DISCUSSION THREAD

**Discussion by anishshriram** [mlcommons/GaNDLF#322] [Q&A]
**GaNDLF Version**
0.0.14-dev

**Desktop**
macOS Monterey
Version 12.1

**How did you install GaNDLF**
```
git clone https://github.com/CBICA/GaNDLF.git
cd GaNDLF
conda create -n venv_gandlf python=3.7 -y
conda activate venv_gandlf
conda install -c conda-forge mamba -y
mamba install -c pytorch pytorch torchvision -y
pip install -e .
```

**Dataset description**
The dataset is titled RSNA_ASNR_MICCAI_BraTS2021_TrainingData_16July2021, referring to the 2021 BraTS training data.

Description from website:
Routine clinically-acquired multi-parametric MRI (mpMRI) scans of glioma, with pathologically confirmed diagnosis and available MGMT promoter methylation status (for the glioblastoma cases with such associated data). All BraTS mpMRI scans are available as NIfTI files (.nii.gz) for Segmentation, which is the use in this issue. These mpMRI scans describe a) native (T1) and b) post-contrast T1-weighted (T1Gd), c) T2-weighted (T2), and d) T2 Fluid Attenuated Inversion Recovery (T2-FLAIR) volumes.

**Describe your question/problem**
I am trying to use gandlf_run.

I then executed it (with specific paths) as so:
```
python gandlf_run \
-c /cbica/home/shrirama/comp_space/project/meningioma/GaNDLF/samples/config_segmentation_brats.yaml \
-i /cbica/home/shrirama/comp_space/project/meningioma/data_processed.csv \
-m /cbica/home/shrirama/comp_space/project/meningioma/models \
-t True \
-d cuda \
-r True
```
The other error I get, though I don't exactly know if it is an error:
```
Please set the environment variable 'CUDA_VISIBLE_DEVICES' correctly before trying to run GANDLF on GPU
```
I assume this has something to do with the sge_wrapper file. I can't find in the wiki information on how to change this variable to the correct settings.

I also moved the sge_wrapper file from samples one directory up. 

I would appreciate some help on this thank you so much.

**Comment by sarthakpati** [comment]
Please post a separate query per question, so that it is easier for us to address.

1. This is related to the file not getting found.
2. Setting the environment variable `CUDA_VISIBLE_DEVICES` is required when trying to run any GPU compute job. See a related post [[ref](https://developer.nvidia.com/blog/cuda-pro-tip-control-gpu-visibility-cuda_visible_devices/)] for details.

**Reply by anishshriram** [reply]
Hello, Sorry I did not see that sooner. I will post separately now.

**Reply by anishshriram** [reply]
But doesn't the 'sge_wrapper' file already properly initialize the `CUDA_VISIBLE_DEVICES` variable? Or do I need to edit that file to set it properly?

**Reply by sarthakpati** [reply]
Are you submitting this job on the CUBIC cluster? If yes, then you might need to do this before calling `gandlf_run` on the compute node (you will need to confirm this with Mark):

```bash
CUDA_VISIBLE_DEVICES=`get_CUDA_VISIBLE_DEVICES` || exit
export CUDA_VISIBLE_DEVICES
```

**Reply by anishshriram** [reply]
I believe that `gandlf_run` submits the job automatically and should do this too? I will confirm with Mr. Bergman.

**Reply by sarthakpati** [reply]
> I believe that `gandlf_run` submits the job automatically

Depends on what you are trying to do with it. If you have set `parallel_compute_command` in the configuration, GaNDLF will try to run the training job as requested. It is the users' duty to ensure the command and environment is set correctly.

**Reply by anishshriram** [reply]
I disabled the `parallel_compute_command` in the configuration as per your previous reccomendation.

**Reply by sarthakpati** [reply]
If it is disabled, GaNDLF doesn't submit any job, rather it runs on the compute node as the user has requested.

**Reply by anishshriram** [reply]
I see. So this means I should in fact set the variable before calling `gandlf_run'

May I ask how I should do this?

**Reply by sarthakpati** [reply]
I have said this in this response: https://github.com/CBICA/GaNDLF/discussions/322#discussioncomment-2265690

**Reply by sarthakpati** [reply]
Great, all the best.

**Reply by anishshriram** [reply]
What I did was call:

```
qsub sge_wrapper
```
in the command line. It did not work. How should I do this properly?

**Reply by sarthakpati** [reply]
This is not a GaNDLF question. Please post this to the CUBIC system admins for clarification.

**Reply by anishshriram** [reply]
I have emailed them thank you.
